source/common/http/mixed_conn_pool.cc

```diff
@@ -0,0 +1,74 @@
+#include "common/http/mixed_conn_pool.h"
+
+#include "common/http/codec_client.h"
+#include "common/http/http1/conn_pool.h"
+#include "common/http/http2/conn_pool.h"
+#include "common/http/utility.h"
+#include "common/tcp/conn_pool.h"
+
+namespace Envoy {
+namespace Http {
+
+Envoy::ConnectionPool::ActiveClientPtr HttpConnPoolImplMixed::instantiateActiveClient() {
+  return std::make_unique<Tcp::ActiveTcpClient>(*this,
+                                                Envoy::ConnectionPool::ConnPoolImplBase::host(), 1);
+}
+
+CodecClientPtr
+HttpConnPoolImplMixed::createCodecClient(Upstream::Host::CreateConnectionData& data) {
+  auto protocol =
+      protocol_ == Protocol::Http11 ? CodecClient::Type::HTTP1 : CodecClient::Type::HTTP2;
+  CodecClientPtr codec{new CodecClientProd(protocol, std::move(data.connection_),
+                                           data.host_description_, dispatcher_, random_generator_)};
+  return codec;
+}
+
+void HttpConnPoolImplMixed::onConnected(Envoy::ConnectionPool::ActiveClient& client) {
+  // When we upgrade from a TCP client to non-TCP we get a spurious onConnected
+  // from the new client. Ignore that.
+  if (client.protocol() != absl::nullopt) {
+    return;
+  }
+
+  connected_ = true;
+  // If an old TLS stack does not negotiate alpn, it likely does not support
+  // HTTP/2. Fail over to HTTP/1.
+  protocol_ = Protocol::Http11;
+  auto tcp_client = static_cast<Tcp::ActiveTcpClient*>(&client);
```

Comment:
Should there by a dynamic_cast ASSERT here? Is there any way to avoid the effective dynamic_cast? Seems like this could be an interface function of some type?